### PR TITLE
CyclicWriteBehindQueue minor improvements

### PR DESCRIPTION
**Pull request by ahmetmircik**
- Added mutable integer
- Fixed wrong method comment.

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/util/MutableInteger.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.util;
+
+/**
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/util/MutableInteger.java]: Can you also document its thread safety (or lack off). It can be perfectly fine that this class isn't save, but make sure everybody knows it.

File: hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CyclicWriteBehindQueue.java [hunk 2]
```diff
@@ -50,11 +51,11 @@
      *
      * @see WriteBehindStore#loadable(com.hazelcast.nio.serialization.Data)
      */
-    private final Map<Data, Integer> index;
+    private final Map<Data, MutableInteger> index;
 
     public CyclicWriteBehindQueue() {
         this.deque = new ArrayDeque<DelayedEntry>();
-        this.index = new HashMap<Data, Integer>();
+        this.index = new HashMap<Data, MutableInteger>();
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CyclicWriteBehindQueue.java]: Why don't you call MultableInteger counter? Makes the code a lot easier to read because the class name now explains its purpose.

File: hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CyclicWriteBehindQueue.java [hunk 3]
```diff
@@ -180,13 +181,13 @@ public void getFrontByNumber(int numberOfElements, Collection<DelayedEntry> coll
...
+        Map<Data, MutableInteger> index = this.index;
 
-        Integer count = index.get(key);
+        MutableInteger count = index.get(key);
         if (count == null) {
-            count = 0;
+            count = new MutableInteger();
         }
-        count++;
+        count.value++;
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CyclicWriteBehindQueue.java]: I would add an inc method on count so you get:

counter.inc()

File: hazelcast/src/main/java/com/hazelcast/util/MutableInteger.java (hunk 1 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/util/MutableInteger.java]: I would place this class in your map.impl package. No need to share it with the rest of the world.

File: hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CyclicWriteBehindQueue.java (hunk 3 above)
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CyclicWriteBehindQueue.java]: what about encapsulating the Map + MutableInteger in something like
```java
public class Counter<T> {
  int incAndGet(T data) {
    [...]
  }  

  int decAntGet(T data) {
    [...]
  }

  [ ...other methods... ]
}
```

this would allow you to remove implementation noise such as
```java
        MutableInteger count = index.get(key);
        if (count == null) {
            count = new MutableInteger();
        }
```
from the CyclicWriteBehindQueue class

File: hazelcast/src/main/java/com/hazelcast/util/MutableInteger.java (hunk 1 above)
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/util/MutableInteger.java]: Added thread-safety comment. I put the class in util because will use it in other modules.

File: hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CyclicWriteBehindQueue.java (hunk 2 above)
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CyclicWriteBehindQueue.java]: will use it in other modules for different purposes.

File: hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CyclicWriteBehindQueue.java (hunk 3 above)
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/CyclicWriteBehindQueue.java]: @pveentjer thought it but this seems very easy to use.
@jerrinot thanks. it will be better to change it in another PR. Only wanted to add `MutableInteger` in this PR.

**Comment by bwzhang2011** [comment]
why not use AtomicInteger ?

**Comment by Donnerbart** [comment]
An `AtomicInteger` introduces synchronization costs. If we don't need thread-safety we shouldn't use concurrent classes to avoid this overhead.

**Comment by bwzhang2011** [comment]
@Donnerbart , you're right. as I just deem it to be global variable so as usual it's better to be set as some Atomic way for processing. so if we don't need thread-safety, just the original prototype is enough.
